lib/widgets/login.dart

```diff
@@ -14,6 +15,85 @@ class _LoginSequenceRoute extends MaterialPageRoute<void> {
   });
 }
 
+enum ServerUrlValidationError {
+  empty,
+  invalidUrl,
+  noUseEmail,
+  unsupportedSchemeZulip,
+  unsupportedSchemeOther;
+
+  /// Whether to wait until the user presses "submit" to give error feedback.
+  ///
+  /// True for errors that will often happen when the user just hasn't finished
+  /// typing a good URL. False for errors that strongly signal a wrong path was
+  /// taken, like when we recognize the form of an email address.
+  bool shouldDeferFeedback() {
+    switch (this) {
+      case empty:
+      case invalidUrl:
+        return true;
+      case noUseEmail:
+      case unsupportedSchemeZulip:
+      case unsupportedSchemeOther:
+        return false;
+    }
+  }
+
+  String message() { // TODO(i18n)
+    switch (this) {
+      case empty:
+        return 'Please enter a URL.';
+      case invalidUrl:
+        return 'Please enter a valid URL.';
+      case noUseEmail:
+        return 'Please enter the server URL, not your email.';
+      case unsupportedSchemeZulip:
+      case unsupportedSchemeOther:
+        return 'The server URL must start with http:// or https://.';
+    }
+  }
+}
+
+class ServerUrlParseResult {
+  ServerUrlParseResult({required this.url, required this.error});
+
+  final Uri? url;
+  final ServerUrlValidationError? error;
+}
+
+class ServerUrlTextEditingController extends TextEditingController {
+  ServerUrlParseResult tryParse() {
+    final trimmedText = text.trim();
+
+    if (trimmedText.isEmpty) {
+      return ServerUrlParseResult(url: null, error: ServerUrlValidationError.empty);
+    }
+
+    Uri? url = Uri.tryParse(trimmedText);
+    if (!RegExp(r'^https?://').hasMatch(trimmedText)) {
+      if (url != null && url.hasScheme && url.scheme == 'zulip') {
```

Comment:
I don't love this logic — the string-matching with a regexp (when we have a perfectly respectable URL parser right there, which we're even also using), and particularly the alternation of parse, then string-match, then use parse result.

But I see that it matches well what we have in zulip-mobile, and that we have no tests for it there.

So I think matching that logic — and adding tests, hooray! — is the right first step. Then, armed with tests, we can refactor and adjust the logic in some later PR.

Reply:
For some valid values of `trimmedText`, we shouldn't be interested in the result of URL parsing at all: in particular, if it's an "https://" URL with the "https://" part left out because it's implicit. For those values, the thing we want URL-parsed is `trimmedText` plus the "https://" prefix.

When `trimmedText` is 'chat.zulip.org', this code is correctly indifferent to `Uri.tryParse(trimmedText)` (which sports features like `.isAbsolute` being false), just as zulip-mobile is correctly indifferent to `tryParseUrl('chat.zulip.org')` (which is undefined).

Possibly there's a better way to write the code with that in mind. Ideally, we'd have a test case or two confirming, for the relevant values of `trimmedText`, that it isn't fooled by whatever we'd get from `Uri.tryParse(trimmedText)`. Can you think of a good test case? Ideally we'd have one, before refactoring… With the help of DartPad, I'm not yet thinking of a `trimmedText` where `Uri.tryParse(trimmedText)` gives null but `Uri.tryParse('https://$trimmedText')` doesn't, but that seems like it would help.

Reply:
Well, as I said, I'd like to get this version merged first before we refactor.

But one observation in particular is that when this inner condition `url != null && url.hasScheme && url.scheme == 'zulip'` is true, then the outer condition `!RegExp(r'^https?://').hasMatch(trimmedText)` will always have been true — if the text began with e.g. `https://`, then it definitely won't have parsed as a URL with scheme `zulip`.

That means that we can swap this condition to come before we ever apply string-matching. Then I think there are some further improvements to make on that path.